metagraph/src/graph/annotated_graph_algorithm.cpp

```diff
@@ -1,264 +1,495 @@
+        if (begin == in_mask.size())
+            return {};
 
-            } else if (labels_out_enc.find(pair.first) != labels_out_enc.end()) {
-                // early cutoff
-                if ((out_count += pair.second) > out_count_cutoff)
-                    return false;
+        size_t end = prev_bit(in_mask, in_mask.size() - 1) + 1;
+        assert(end > begin);
 
-            } else {
-                other_count += pair.second;
+        out_kmer_count -= count_ones(out_mask, 0, begin) + count_ones(out_mask, end, out_mask.size());
+        size_t other_kmer_count = check_other ? count_ones(other_mask, begin, end) : 0;
+
+        size_t cur_size = end - begin;
+        size_t label_in_cutoff = std::ceil(config.label_mask_in_unitig_fraction * cur_size);
+        size_t label_out_cutoff = std::floor(config.label_mask_out_unitig_fraction * cur_size);
+        size_t other_cutoff = std::floor(config.label_mask_other_unitig_fraction * cur_size);
+
+        if (in_kmer_count >= label_in_cutoff && out_kmer_count <= label_out_cutoff
+                && other_kmer_count <= other_cutoff)
+            return { std::make_pair(begin, end) };
+
+
+        return {};
+    };
+
+    update_masked_graph_by_unitig(*masked_graph, get_kept_intervals, num_threads);
+
+    return std::move(*masked_graph);
+}
+
+std::shared_ptr<MaskedDeBruijnGraph>
+make_initial_masked_graph(std::shared_ptr<const DeBruijnGraph> graph_ptr,
+                          sdsl::int_vector<> &counts,
+                          std::unique_ptr<bitmap>&& mask,
+                          bool add_complement,
+                          bool make_boss,
+                          size_t num_threads) {
+    // counts is a double-length vector storing the in-label and out-label
+    // counts interleaved
+    assert(counts.size() == mask->size() * 2);
+
+    graph_ptr = std::shared_ptr<const DeBruijnGraph>(
+        std::shared_ptr<const DeBruijnGraph>(), &graph_ptr->get_base_graph()
+    );
+
+    auto masked_graph = std::make_shared<MaskedDeBruijnGraph>(
+        graph_ptr,
+        std::move(mask),
+        true,
+        graph_ptr->get_mode()
+    );
+
+    logger->trace("Constructed masked graph with {} nodes", masked_graph->num_nodes());
+
+    bool masked_canonical = add_complement
+        || graph_ptr->get_mode() == DeBruijnGraph::CANONICAL;
+
+    if (make_boss
+            || (add_complement && graph_ptr->get_mode() != DeBruijnGraph::CANONICAL)) {
+        // we can't guarantee that the reverse complement is present, so
+        // construct a new subgraph
+        logger->trace("Constructing BOSS from labeled subgraph");
+        std::vector<std::pair<std::string, sdsl::int_vector<>>> contigs;
+        std::mutex add_mutex;
+        BOSSConstructor constructor(
+            graph_ptr->get_k() - 1,
+            masked_canonical,
+            0 /* count width */, "" /* suffix */, 1 /* num_threads */,
+            masked_graph->num_nodes() * 32
+        );
```

Comment:
Why do you need to rebuild a boss graph?
What is the workflow? (add a comment above to describe what this function `make_initial_masked_graph` is doing)
Can't this be done separately in the `build` mode?

Reply:
I reworked the process so it no longer needs a new BOSS graph to be built.